lib/puppet/pops/validation/checker4_0.rb

```diff
@@ -530,6 +533,101 @@ def internal_check_future_reserved_word(o, name)
     end
   end
 
+  def internal_check_file_namespace(o, name, file)
+    return if file.nil? || file == '' #e.g. puppet apply -e '...'
+
+    lc_file = file.downcase
+
+    file_namespace = @file_to_namespace[lc_file]
+    if file_namespace.nil?
+      return if @file_to_namespace.key?(lc_file)
+
+      file_namespace = @file_to_namespace[lc_file] = namespace_for_file(lc_file)
+      return if file_namespace.nil?
+    end
+
+    if !name.downcase.start_with?(file_namespace)
+      acceptor.accept(Issues::ILLEGAL_DEFINITION_LOCATION, o, {:name => name, :file => file})
+    end
+  end
+
+  NEVER_MATCH = '\b'.freeze
+
+  def namespace_for_file(file)
+    path = Pathname.new(file)
+
+    return nil if path.extname != ".pp"
+
+    #We want an expanded path, but lowercase, so...
+    path = Pathname.new(path.expand_path.to_s.downcase)
+
+    env = Puppet.lookup(:current_environment)
+    return nil if env.nil? || initial_manifest?(path, env)
+
+    #All auto-loaded files from modules come from a module search path dir
+    relative_path = get_module_relative_path(path, env)
+
+    return nil if relative_path.nil?
+
+    #If a file comes from a module, but isn't in the right place, always error
+    names = dir_to_names(relative_path)
+    return NEVER_MATCH if names.nil?
+
+    names.join("::").freeze
+  end
+
+  def initial_manifest?(path, env)
+    manifest_setting = env.manifest
```

Comment:
Should this be `env.manifest` or `env.configuration.manifest`? Eg when running `puppet apply manifest.pp` I see:

```
(byebug) env.manifest
"/Users/josh/work/puppet/manifest.pp"
(byebug) env.configuration.manifest
:no_manifest
```

Reply:
Using `env.manifest` is correct